tests/drivers/sensor/bmi160/src/i2c.cpp

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/emul_sensor.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/ztest.h>
+
+#include "bmi160.h"
+#include "checks.hpp"
+#include "fixture.hpp"
+
+ZTEST_USER_F(bmi160, test_bmi160_i2c_get_offset_fail_to_read_offset_acc)
+{
+	static int fail_reg_number = -1;
+	struct i2c_emul_api mock_bus_api;
+	struct sensor_value value;
+
+	fixture->emul_i2c->bus.i2c->mock_api = &mock_bus_api;
+	mock_bus_api.transfer = [](const struct emul *, struct i2c_msg *msgs, int num_msgs, int) {
+		if (fail_reg_number >= 0 && i2c_is_read_op(&msgs[1]) &&
+		    bmi160_i2c_is_touching_reg(msgs, num_msgs, fail_reg_number)) {
+			return -EIO;
+		}
+		return -ENOSYS;
+	};
```

Comment:
@teburd this is the reason for C++. I can use a lambda here which makes the test much easier to read since the mock is inside the test function.

Reply:
https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html#enabling-c-support

The docs promise C++ support in application code. I don't see any notes about testing code. I generally find C++ incredibly difficult to read but I'm not the sole developer here, maybe writing tests in C++ would be ok, but thats something that likely needs to be brought up in the arch wg. I guess the next question would be what C++ standard? Seemingly C++98 is the only thing guaranteed and I don't believe lambdas are in 98.

Reply:
I can't find it in the meeting notes (I guess it wasn't an important comment to write down), but I believe we had this discussion a long time ago in the TSC where we said that tests are fine in C++. This came up when I first added CHRE. I know we don't explicitly ban them:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/9535482/4e2fb6b9-50d1-4dea-abb0-e51f3e75c388)

There are more tests that we can enable via C++ class fixtures later along with the emulators. I personally feel like it gives us a lot of power to test things more easily and quickly.

Reply:
I am with @teburd on this one. I do not recall having such a discussion in the TSC and I certainly do not support starting to use C++ in what is otherwise a C project with support for applications written in C++. Sorry.